src/windows/Stack/DefaultColorWhen.hs

```diff
@@ -0,0 +1,83 @@
+{-# LANGUAGE CPP #-}
+
+{- | This version of the module is only for Windows operating systems.
+-}
+module Stack.DefaultColorWhen
+  ( defaultColorWhen
+  ) where
+
+-- The Win32 package provides CPP macro WINDOWS_CCONV.
+#include "windows_cconv.h"
+
+import Stack.Prelude (Exception, SomeException, stdout, try)
+import Stack.Types.Runner (ColorWhen (ColorAuto, ColorNever))
+
+import Control.Exception (throw)
+import Data.Bits ((.|.), (.&.))
+import Data.Typeable (Typeable)
+import Foreign.Marshal (alloca)
+import Foreign.Ptr (Ptr)
+import Foreign.Storable (peek)
+import System.Win32.Types (BOOL, DWORD, ErrCode, HANDLE, getLastError,
+  iNVALID_HANDLE_VALUE, nullHANDLE, withHandleToHANDLE)
+
+defaultColorWhen :: IO ColorWhen
+defaultColorWhen = withHandleToHANDLE stdout aNSISupport
+
+-- The following is based on extracts from the modules
+-- System.Console.ANSI.Windows.Foreign and System.Console.ANSI.Windows.Detect
+-- from the ansi-terminal package.
+
+-- | This function first checks if the Windows handle is valid and yields
+-- 'never' if it is not. It then tries to get a ConHost console mode for that
+-- handle. If it can not, it assumes that the handle is ANSI-enabled. If virtual
+-- termimal (VT) processing is already enabled, the handle supports 'auto'.
+-- Otherwise, it trys to enable processing. If it can, the handle supports
+-- 'auto'. If it can not, the function yields 'never'.
+aNSISupport :: HANDLE -> IO ColorWhen
+aNSISupport h =
+  if h == iNVALID_HANDLE_VALUE || h == nullHANDLE
+    then return ColorNever  -- Invalid handle or no handle
+    else do
+      tryMode <- try (getConsoleMode h) :: IO (Either SomeException DWORD)
```

Comment:
I'd recommend using `tryAny` here: it will make the type signature unnecessary, and be more explicit about using a function that doesn't catch async exceptions.

Reply:
Or did you mean to catch only a `ConsoleException` from below?

Reply:
I've now moved away from duplicating `ansi-terminal`'s use of the exception mechanism for this small extract.